### ISSUE TITLE
ur_kinematics not working with limited joint range

### ISSUE THREAD

**Issue by tnl-woi** [ros-industrial/universal_robot#127]
Since the update in which URKinematicsPlugin is used instead of KDLKinematicsPlugin I have troubles with the motion planning. When I use the URKinematicsPlugin with limited joint range (parameter "limited:=true"), motion planning fails consistently and fast, and the following lines are logged:

```
[ INFO] [1413803613.589846940]: Planning request received for MoveGroup action. Forwarding to planning pipeline.
[ INFO] [1413803613.593838054]: No planner specified. Using default.
[ INFO] [1413803613.594163867]: LBKPIECE1: Attempting to use default projection.
[ INFO] [1413803613.595661696]: LBKPIECE1: Starting with 1 states
[ INFO] [1413803613.711591070]: State outside bounds
[ INFO] [1413803613.711821634]: State outside bounds
[ INFO] [1413803613.711952278]: State outside bounds
[ INFO] [1413803613.712079082]: State outside bounds
[ INFO] [1413803613.712203645]: State outside bounds
[ INFO] [1413803613.712329330]: State outside bounds
[ INFO] [1413803613.712449617]: State outside bounds
[ INFO] [1413803613.712572634]: State outside bounds
[ INFO] [1413803613.712740112]: State outside bounds
[ INFO] [1413803613.712868885]: State outside bounds
[ INFO] [1413803613.712991710]: State outside bounds
[ INFO] [1413803613.713112992]: State outside bounds
[ INFO] [1413803613.713234464]: State outside bounds
[ INFO] [1413803613.713357884]: State outside bounds
[ INFO] [1413803613.713477776]: State outside bounds
[ INFO] [1413803613.713590285]: State outside bounds
[ INFO] [1413803613.713731441]: State outside bounds
[ INFO] [1413803613.713845021]: State outside bounds
[ INFO] [1413803613.713956083]: State outside bounds
[ INFO] [1413803613.714066334]: State outside bounds
[ INFO] [1413803613.714177697]: State outside bounds
[ INFO] [1413803613.714289183]: State outside bounds
[ INFO] [1413803613.714421740]: State outside bounds
[ INFO] [1413803613.714545019]: State outside bounds
[ INFO] [1413803613.714707252]: State outside bounds
[ INFO] [1413803613.714832176]: State outside bounds
[ INFO] [1413803613.714955614]: State outside bounds
[ INFO] [1413803613.715070355]: State outside bounds
[ INFO] [1413803613.715196389]: State outside bounds
[ INFO] [1413803613.715308802]: State outside bounds
[ INFO] [1413803613.715420378]: State outside bounds
[ INFO] [1413803613.715688061]: State outside bounds
[ERROR] [1413803613.827743907]: LBKPIECE1: Unable to sample any valid states for goal tree
[ INFO] [1413803613.827813086]: LBKPIECE1: Created 1 (1 start + 0 goal) states in 1 cells (1 start (1 on boundary) + 0 goal (0 on boundary))
[ INFO] [1413803613.827851327]: No solution found after 0.232906 seconds
[ INFO] [1413803613.827987487]: Unable to solve the planning problem
```

When I use the `URKinematicsPlugin` with normal joint range ("limited:=false") I don't see this messages, but the motion planning still (almost?) always fails. It does generate motion plans, however, but they always seem to collide so they are rejected as a valid plan. This may be another issue: MoveIt! is known to have problems with the full joint range of the UR5, and there are some open issues with collision detection that might affect this.

The fact that it logs "State outside bounds" when using limited joint range should give some hints about the cause of the problem. I checked that all joints of the robot in the start state are within [-PI, PI], so that's not the problem. I wonder if the ur_kinematics module still honors the limited joint range.

**Comment by fmessmer** [comment]
In #128, the KDLKinematics plugin is used as default again!

@kphawkins 
Is it possible to generate a URXKinematicsPlugin that is able to be used with both versions - limited and unlimited URX?
Or do we need a URXLimitedKinematicsPlugin for the UR5 and UR10 respectively?

**Comment by fmessmer** [comment]
Yet another effect of this problem can be seen when the interactive marker in rviz is used to set a goal configuration for the robot. It's almost impossible to drag the marker to a config that does not violate joint limits - as reported [here](https://groups.google.com/forum/#!topic/swri-ros-pkg-dev/jFMKM0_xga8)

One more reason to use KDLKinematics plugin as default again (see #128)

**Comment by fmessmer** [comment]
in #128 the KDLKinematics plugin has been made default again...
Can we close this issue? @tnl-woi

**Comment by tnl-woi** [comment]
@ipa-fxm, I think the question remains:

> @kphawkins
> Is it possible to generate a URXKinematicsPlugin that is able to be used with both versions - limited and unlimited URX?

The default settings indeed work fine now, which is a Good Thing, but I was under the impression that the URKinematicsPlugin is better (faster?) than the KDLKinematicsPlugin. However, it cannot be used (with MoveIt!) in its current form. I think this is a shortcoming of (or at least a feature request for :p) the universal_driver package, so I'd propose to leave this issue open until a working URx-specific kinematics plugin is available.

But you might consider this a MoveIt! issue, which would be a valid reason to close this issue anyway.
I'll leave the open/close decision up to you, if you've taken this arguments into account ;-)

**Comment by fmessmer** [comment]
Ok, thanks for the feedback!
But I think it's not a MoveIt! issue as MoveIt! is doing nothing wrong, but rejects IK solutions from URKintematicsPlugin because the solutions are for an unlimited UR and MoveIt! (and URDF) uses limited joint ranges...

Thus, indeed the question remains:
@kphawkins
Is it possible to generate a URXKinematicsPlugin that is able to be used with both versions - limited and unlimited URX?

**Comment by fmessmer** [comment]
Please test my [test branch](https://github.com/ipa-fxm/universal_robot/tree/investigate_ur_kinematics)...

**Comment by fmessmer** [comment]
#184 fixes this...

**Comment by kphawkins** [comment]
Nice work!

**Comment by abubeck** [comment]
Closed with #184
